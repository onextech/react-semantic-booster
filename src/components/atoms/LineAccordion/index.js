import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Accordion, Icon } from 'semantic-ui-react';

const sizes = ['mini', 'tiny', 'small', 'medium', 'large', 'big', 'huge', 'massive'];

const StyledAccordion = styled(Accordion)`
  &.ui.accordion {
    .container {
      border-bottom: 1px solid rgba(0,0,0,0.2);
    }
    .title {
      display: flex;
      justify-content: space-between;
      align-items: center;
      &:hover, &:focus, &:active {
        font-weight: bold;
      }
    }
    
    &.inverted {
      .container {
        border-bottom: 1px solid white;
      }
      .title {
        opacity: 0.7;
        &:hover, &:focus, &:active {
          opacity: 1;
          font-weight: normal;
        }
      }
    }
    
  ${(props) => {
    if (props.size) {
      const found = sizes.indexOf(props.size);
      return found && `
        .title {
          padding: ${(found + 0.5) * 0.5}em 0 ${(found + 0.3) * 0.5}em;
          font-size: ${(found + 0.5) * 0.5}em;
        }
        .content {
          padding-bottom: ${(found + 0.5) * 0.5}em !important;
        }   
      `;
    }
    return false;
  }}
}`;

class LineAccordion extends React.Component {
  state = { activeIndex: this.props.defaultindex }

  getAccordionClassName = () => {
    const className = [];
    const { size } = this.props;
    if (size) {
      className.push(size);
    }
    return className.join(' ');
  }

  handleClick = (e, titleProps) => {
    const { index } = titleProps;
    const { activeIndex } = this.state;
    const newIndex = activeIndex === index ? -1 : index;
    this.setState({ activeIndex: newIndex });
  }

  render() {
    const { activeIndex } = this.state;
    const { panels, ...accordionProps } = this.props;

    return (
      <StyledAccordion className={this.getAccordionClassName()} {...accordionProps}>
        {
          panels.map((panel, i) => (
            <div key={i} className="container">
              <Accordion.Title active={activeIndex === i} index={i} onClick={this.handleClick}>
                {panel.title.content}
                <Icon name="plus" />
              </Accordion.Title>
              <Accordion.Content active={activeIndex === i}>
                {panel.content.content}
              </Accordion.Content>
            </div>
          ))
        }
      </StyledAccordion>
    );
  }
}

LineAccordion.propTypes = {
  panels: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.shape({
      content: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.string,
        PropTypes.array,
      ]),
    }),
    content: PropTypes.shape({
      content: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.string,
        PropTypes.array,
      ]),
    }),
  })).isRequired,
  size: PropTypes.string, // enum: size constant above
  defaultindex: PropTypes.number, // zero-based index
};

LineAccordion.defaultProps = {
  size: undefined,
  defaultindex: undefined,
};

export default LineAccordion;
