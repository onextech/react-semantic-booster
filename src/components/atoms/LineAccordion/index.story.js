import React from 'react';
import { storiesOf } from '@storybook/react';
import { Segment } from 'semantic-ui-react';
import LineAccordion from '.';

// eslint-disable-next-line import/prefer-default-export
export const panelsExample = [
  {
    title: {
      content: 'Residencies',
    },
    content: {
      content: <ul>
        <li>Resident DJ - Peaches Club, Singapore (former Cherry Disco) (2017 - 2018)</li>
        <li>Resident DJ - PLAY Club since 2010</li>
      </ul>,
    },
  },
  {
    title: {
      content: 'Venues Played',
    },
    content: {
      content: <ul>
        <li>Attica Club, Singapore</li>
        <li>Avalon Club, Singapore</li>
        <li>Filter Club, Singapore</li>
        <li>New Asia Bar, Singapore</li>
      </ul>,
    },
  },
];

storiesOf('LineAccordion', module)
  .add('Default', () => (
    <div>
      <Segment>
        <LineAccordion fluid size="small" panels={panelsExample} />
      </Segment>
      <Segment inverted>
        <LineAccordion inverted fluid defaultindex={1} panels={panelsExample} />
      </Segment>
    </div>
  ));

