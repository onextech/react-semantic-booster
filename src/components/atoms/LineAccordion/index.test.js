import React from 'react';
import renderer from 'react-test-renderer';
import LineAccordion from '.';
import { panelsExample } from './index.story';


describe('<LineAccordion>', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<LineAccordion panels={panelsExample} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
