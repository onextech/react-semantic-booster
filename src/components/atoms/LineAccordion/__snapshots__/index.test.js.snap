// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LineAccordion> renders correctly 1`] = `
<div
  className="accordion ui LineAccordion__StyledAccordion-jXXxAU hBYvMi"
  defaultindex={undefined}
  size={undefined}
>
  <div
    className="container"
  >
    <div
      className="title"
      onClick={[Function]}
    >
      Residencies
      <i
        aria-hidden="true"
        className="plus icon"
      />
    </div>
    <div
      className="content"
    >
      <ul>
        <li>
          Resident DJ - Peaches Club, Singapore (former Cherry Disco) (2017 - 2018)
        </li>
        <li>
          Resident DJ - PLAY Club since 2010
        </li>
      </ul>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="title"
      onClick={[Function]}
    >
      Venues Played
      <i
        aria-hidden="true"
        className="plus icon"
      />
    </div>
    <div
      className="content"
    >
      <ul>
        <li>
          Attica Club, Singapore
        </li>
        <li>
          Avalon Club, Singapore
        </li>
        <li>
          Filter Club, Singapore
        </li>
        <li>
          New Asia Bar, Singapore
        </li>
      </ul>
    </div>
  </div>
</div>
`;
